router_client.go

```diff
@@ -560,3 +649,209 @@ func (r *routerClient) SubscribeHtlcEvents(ctx context.Context) (
 
 	return htlcChan, errChan, nil
 }
+
+// InterceptHtlcs intercepts htlcs on a node, using the handler function
+// provided to provide the interceptor with interception decisions. The handler
+// provided may block, but must exit if the context passed in is canceled. It
+// is also responsible for releasing htlcs on exit.
+//
+// There are a few ways in which this method can exit:
+// - ctx canceled: the calling client cancels
+// - r.quit: the router is shut down
+// - lnd stream error: lnd has exited
+// - handler error: a critical error occurred while handing a htlc
+func (r *routerClient) InterceptHtlcs(ctx context.Context,
+	handler HtlcInterceptHandler) error {
+
+	// Create a child context that will be canceled when this function
+	// exits. We use this context to be able to cancel goroutines when we
+	// exit on errors, because the parent context won't be canceled in that
+	// case.
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	stream, err := r.client.HtlcInterceptor(
+		r.routerKitMac.WithMacaroonAuth(ctx),
+	)
+	if err != nil {
+		return err
+	}
+
+	// Create an error channel that we'll send errors on if any of our
+	// goroutines fail. We buffer by 1 so that the goroutine doesn't depend
+	// on the stream being read, and select on context cancelation and
+	// quit channel so that we do not block in the case where we exit with
+	// multiple errors.
+	errChan := make(chan error, 1)
+
+	sendErr := func(err error) {
+		select {
+		case errChan <- err:
+		case <-ctx.Done():
+		case <-r.quit:
+		}
+	}
+
+	// Start a goroutine that consumes interception requests from lnd and
+	// sends them into our requests channel for handling. The requests
+	// channel is not buffered because we expect all requests to be handled
+	// until this function exits, at which point we expect our context to
+	// be canceled or quit channel to be closed.
+	requestChan := make(chan InterceptedHtlc)
+	r.wg.Add(1)
+	go func() {
+		defer r.wg.Done()
+
+		for {
+			// Do a quick check whether our client context has been
+			// canceled so that we can exit sooner if needed.
+			if ctx.Err() != nil {
```

Comment:
If the client context has been canceled, shouldn't the below stream return an error?

Reply:
I _think_ that it only returns the error once the server tries to send a new item (was testing this locally and noticed that it sometimes took a little time to exit when this ctx was canceled). Could be wrong there though. Adding this just speeds us up a little. Will do dig around the grpc code, because I'm also curious about it.